rpc/lib/client/ws_client.go

```diff
@@ -1,160 +1,390 @@
 package rpcclient
 
 import (
+	"context"
 	"encoding/json"
+	"fmt"
+	"math"
 	"net"
 	"net/http"
+	"sync"
 	"time"
 
 	"github.com/gorilla/websocket"
 	"github.com/pkg/errors"
+	metrics "github.com/rcrowley/go-metrics"
+
 	types "github.com/tendermint/tendermint/rpc/lib/types"
 	cmn "github.com/tendermint/tmlibs/common"
 )
 
 const (
-	wsResultsChannelCapacity = 10
-	wsErrorsChannelCapacity  = 1
-	wsWriteTimeoutSeconds    = 10
+	// Time allowed to write a message to the peer.
+	writeWait = 10 * time.Second
+
+	// Maximum reconnect attempts
+	maxReconnectAttempts = 25
+
+	defaultPongWait   = 30 * time.Second
+	defaultPingPeriod = (defaultPongWait * 9) / 10
 )
 
 type WSClient struct {
 	cmn.BaseService
+
+	conn *websocket.Conn
+
 	Address  string // IP:PORT or /path/to/socket
 	Endpoint string // /websocket/url/endpoint
 	Dialer   func(string, string) (net.Conn, error)
-	*websocket.Conn
-	ResultsCh chan json.RawMessage // closes upon WSClient.Stop()
-	ErrorsCh  chan error           // closes upon WSClient.Stop()
+
+	PingPongLatencyTimer metrics.Timer
+	sentLastPingAt       time.Time
+
+	// user facing channels, closed only when the client is being stopped.
+	ResultsCh chan json.RawMessage
+	ErrorsCh  chan error
+
+	// internal channels
+	send               chan types.RPCRequest // user requests
+	backlog            chan types.RPCRequest // stores a single user request received during a conn failure
+	reconnectAfter     chan error            // reconnect requests
+	receiveRoutineQuit chan struct{}         // a way for receiveRoutine to close writeRoutine
+
+	reconnecting bool
+
+	wg  sync.WaitGroup
+	mtx sync.RWMutex
+
+	// Time allowed to read the next pong message from the server.
+	pongWait time.Duration
+
+	// Send pings to server with this period. Must be less than pongWait.
+	pingPeriod time.Duration
 }
 
-// create a new connection
-func NewWSClient(remoteAddr, endpoint string) *WSClient {
+// NewWSClient returns a new client.
+func NewWSClient(remoteAddr, endpoint string, options ...func(*WSClient)) *WSClient {
 	addr, dialer := makeHTTPDialer(remoteAddr)
-	wsClient := &WSClient{
-		Address:  addr,
-		Dialer:   dialer,
-		Endpoint: endpoint,
-		Conn:     nil,
+	c := &WSClient{
+		Address:              addr,
+		Dialer:               dialer,
+		Endpoint:             endpoint,
```

Comment:
like for people who may use this library? we use `"/websocket"` everywhere. don't think we need a validation for that.